SameLen property should be inferred on both variables

Consider the following code:
`import org.checkerframework.checker.index.qual.*;`
`class checkerr{`
`public void abc(String x, @SameLen("#1") String y){`
`}`
`public void err(@SameLen("#2") String a, String b){`
`abc(a,b);`
`}`
`}`
Run it by typing in the terminal:
javac -processor org.checkerframework.checker.index.IndexChecker checkerr.java

Output:
checkerr.java:6: error: [argument.type.incompatible] incompatible types in argument.
abc(a,b);
      ^
  found   : @SameLen("b") String
  required: @SameLen("a") String
1 error

Now if we annotate 'a' to be of the same length as of 'b', 'b' should automatically be inferred to have the same length as 'a', which clearly fails during above code.

Thanks for noticing this bug!

Here is an expanded test case, which I have committed to the repository:

```
import org.checkerframework.checker.index.qual.*;

class SameLenOnFormalParameter {
  public void requiresSameLen1(String x, @SameLen("#1") String y) {}

  public void requiresSameLen2(@SameLen("#2") String x, String y) {}

  public void m1(@SameLen("#2") String a, String b) {
    requiresSameLen1(a, b);
  }

  public void m2(@SameLen("#2") String a, String b) {
    @SameLen("a") String b2 = b;
    requiresSameLen1(a, b2);
  }

  public void m3(@SameLen("#2") String a, String b) {
    @SameLen("b") String a2 = a;
    @SameLen("a") String b2 = b;
    requiresSameLen1(a, b2);
  }

  public void m4(@SameLen("#2") String a, String b) {
    requiresSameLen2(a, b);
  }
}
```

and here is the ouput:

```
SameLenOnFormalParameter.java:9: error: [argument.type.incompatible] incompatible types in argument.
    requiresSameLen1(a, b);
                        ^
  found   : @SameLen("b") String
  required: @SameLen("a") String
SameLenOnFormalParameter.java:13: error: [assignment.type.incompatible] incompatible types in assignment.
    @SameLen("a") String b2 = b;
                              ^
  found   : @SameLen("b") String
  required: @SameLen("a") String
```

Notice the surprising fact that `m2` fails but `m3` succeeds, although `m3` has merely added an irrelevant, usused assignment.  The reason is that at assignments, the SameLen Checker merges annotations:  it gives both the left-hand side and the right-hand side the same annotations.

The SameLen Checker should do that same merging at procedure entries.

PS: Please [use "```" to quote multi-line code](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown).  That also retains indentation.  (Compare the look of your code example with mine.)

--- a/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
@@ -266,6 +266,7 @@ public class SameLenTransfer extends CFTransfer {
             UnderlyingAST.CFGMethod method,
             MethodTree methodTree,
             ExecutableElement methodElement) {
+        super.addInformationFromPreconditions(info, factory, method, methodTree, methodElement);
         List<? extends VariableTree> paramTrees = methodTree.getParameters();
 
         List<String> paramNames =

--- a/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
@@ -6,7 +6,6 @@ import com.sun.source.util.TreePath;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
-import java.util.stream.Collectors;
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.type.TypeKind;
@@ -267,29 +266,26 @@ public class SameLenTransfer extends CFTransfer {
             MethodTree methodTree,
             ExecutableElement methodElement) {
         super.addInformationFromPreconditions(info, factory, method, methodTree, methodElement);
+
         List<? extends VariableTree> paramTrees = methodTree.getParameters();
+        List<String> paramNames = new ArrayList<>();
+        List<AnnotatedTypeMirror> params = new ArrayList<>();
 
-        List<String> paramNames =
-                paramTrees.stream()
-                        .map(t -> ((VariableTree) t).getName().toString())
-                        .collect(Collectors.toList());
+        for (VariableTree tree : paramTrees) {
+            paramNames.add(tree.getName().toString());
+            params.add(aTypeFactory.getAnnotatedType(tree));
+        }
 
-        List<AnnotatedTypeMirror> params =
-                paramTrees.stream().map(factory::getAnnotatedType).collect(Collectors.toList());
-
-        // for each parameter,
         for (int index = 0; index < params.size(); index++) {
 
-            // get the type of the parameter
+            // if the parameter has a samelen annotation, then look
+            // for other parameters in that annotation and propagate
+            // default the other annotation so that it is symmetric
             AnnotatedTypeMirror atm = params.get(index);
-            // check to see if it has a @SameLen annotation
             AnnotationMirror anm = atm.getAnnotation(SameLen.class);
             if (anm != null) {
-                // If it has, then get the list of arguments from @SameLen
                 List<String> values = IndexUtil.getValueOfAnnotationWithStringArgument(anm);
-                // For each String
                 for (String value : values) {
-                    // check if the value is in the list of parameters
                     int otherParamIndex = paramNames.indexOf(value);
                     if (otherParamIndex != -1) {
                         // the SameLen value is in the list of params, so modify the type of

--- a/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
@@ -302,9 +302,8 @@ public class SameLenTransfer extends CFTransfer {
                 Receiver otherParamRec = null;
                 try {
                     otherParamRec =
-                            aTypeFactory.getReceiverFromJavaExpressionString(
-                                    paramNames.get(otherParamIndex),
-                                    aTypeFactory.getPath(paramTrees.get(otherParamIndex)));
+                            FlowExpressionParseUtil.internalReprOfVariable(
+                                    aTypeFactory, paramTrees.get(otherParamIndex));
                 } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
                     // do nothing
                 }

--- a/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
@@ -283,30 +283,33 @@ public class SameLenTransfer extends CFTransfer {
             // default the other annotation so that it is symmetric
             AnnotatedTypeMirror atm = params.get(index);
             AnnotationMirror anm = atm.getAnnotation(SameLen.class);
-            if (anm != null) {
-                List<String> values = IndexUtil.getValueOfAnnotationWithStringArgument(anm);
-                for (String value : values) {
-                    int otherParamIndex = paramNames.indexOf(value);
-                    if (otherParamIndex != -1) {
-                        // the SameLen value is in the list of params, so modify the type of
-                        // that param in the store
-                        AnnotationMirror newSameLen =
-                                aTypeFactory.createSameLen(
-                                        Collections.singletonList(paramNames.get(index)));
-                        Receiver otherParamRec = null;
-                        try {
-                            otherParamRec =
-                                    aTypeFactory.getReceiverFromJavaExpressionString(
-                                            paramNames.get(otherParamIndex),
-                                            aTypeFactory.getPath(paramTrees.get(otherParamIndex)));
-                        } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
-                            // do nothing
-                        }
+            if (anm == null) {
+                continue;
+            }
 
-                        if (otherParamRec != null) {
-                            info.insertValue(otherParamRec, newSameLen);
-                        }
-                    }
+            List<String> values = IndexUtil.getValueOfAnnotationWithStringArgument(anm);
+            for (String value : values) {
+                int otherParamIndex = paramNames.indexOf(value);
+                if (otherParamIndex == -1) {
+                    continue;
+                }
+
+                // the SameLen value is in the list of params, so modify the type of
+                // that param in the store
+                AnnotationMirror newSameLen =
+                        aTypeFactory.createSameLen(
+                                Collections.singletonList(paramNames.get(index)));
+                Receiver otherParamRec = null;
+                try {
+                    otherParamRec =
+                            aTypeFactory.getReceiverFromJavaExpressionString(
+                                    paramNames.get(otherParamIndex),
+                                    aTypeFactory.getPath(paramTrees.get(otherParamIndex)));
+                } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+                    // do nothing
+                }
+                if (otherParamRec != null) {
+                    info.insertValue(otherParamRec, newSameLen);
                 }
             }
         }

--- a/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/samelen/SameLenTransfer.java
@@ -1,9 +1,14 @@
 package org.checkerframework.checker.index.samelen;
 
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.VariableTree;
 import com.sun.source.util.TreePath;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.stream.Collectors;
 import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.type.TypeKind;
 import org.checkerframework.checker.index.IndexUtil;
 import org.checkerframework.checker.index.qual.SameLen;
@@ -12,6 +17,7 @@ import org.checkerframework.dataflow.analysis.FlowExpressions;
 import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
 import org.checkerframework.dataflow.analysis.TransferInput;
 import org.checkerframework.dataflow.analysis.TransferResult;
+import org.checkerframework.dataflow.cfg.UnderlyingAST;
 import org.checkerframework.dataflow.cfg.node.ArrayCreationNode;
 import org.checkerframework.dataflow.cfg.node.AssignmentNode;
 import org.checkerframework.dataflow.cfg.node.FieldAccessNode;
@@ -21,6 +27,8 @@ import org.checkerframework.framework.flow.CFAnalysis;
 import org.checkerframework.framework.flow.CFStore;
 import org.checkerframework.framework.flow.CFTransfer;
 import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
 import org.checkerframework.framework.util.FlowExpressionParseUtil;
 import org.checkerframework.javacutil.AnnotationUtils;
 
@@ -248,5 +256,62 @@ public class SameLenTransfer extends CFTransfer {
         }
 
         return new ConditionalTransferResult<>(result.getResultValue(), thenStore, elseStore);
+    }
+
+    /** Overridden to ensure that SameLen annotations on method parameters are symmetric. */
+    @Override
+    protected void addInformationFromPreconditions(
+            CFStore info,
+            AnnotatedTypeFactory factory,
+            UnderlyingAST.CFGMethod method,
+            MethodTree methodTree,
+            ExecutableElement methodElement) {
+        List<? extends VariableTree> paramTrees = methodTree.getParameters();
+
+        List<String> paramNames =
+                paramTrees.stream()
+                        .map(t -> ((VariableTree) t).getName().toString())
+                        .collect(Collectors.toList());
+
+        List<AnnotatedTypeMirror> params =
+                paramTrees.stream().map(factory::getAnnotatedType).collect(Collectors.toList());
+
+        // for each parameter,
+        for (int index = 0; index < params.size(); index++) {
+
+            // get the type of the parameter
+            AnnotatedTypeMirror atm = params.get(index);
+            // check to see if it has a @SameLen annotation
+            AnnotationMirror anm = atm.getAnnotation(SameLen.class);
+            if (anm != null) {
+                // If it has, then get the list of arguments from @SameLen
+                List<String> values = IndexUtil.getValueOfAnnotationWithStringArgument(anm);
+                // For each String
+                for (String value : values) {
+                    // check if the value is in the list of parameters
+                    int otherParamIndex = paramNames.indexOf(value);
+                    if (otherParamIndex != -1) {
+                        // the SameLen value is in the list of params, so modify the type of
+                        // that param in the store
+                        AnnotationMirror newSameLen =
+                                aTypeFactory.createSameLen(
+                                        Collections.singletonList(paramNames.get(index)));
+                        Receiver otherParamRec = null;
+                        try {
+                            otherParamRec =
+                                    aTypeFactory.getReceiverFromJavaExpressionString(
+                                            paramNames.get(otherParamIndex),
+                                            aTypeFactory.getPath(paramTrees.get(otherParamIndex)));
+                        } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+                            // do nothing
+                        }
+
+                        if (otherParamRec != null) {
+                            info.insertValue(otherParamRec, newSameLen);
+                        }
+                    }
+                }
+            }
+        }
     }
 }

--- a/checker/tests/index/Issue2493.java
+++ b/checker/tests/index/Issue2493.java
@@ -1,0 +1,9 @@
+// test case for issue 2493: http://tinyurl.com/cfissue/2493
+
+import org.checkerframework.checker.index.qual.*;
+
+class Issue2493 {
+    public static void test(int a[], int @SameLen("#1") [] b) {
+        for (@IndexOrHigh("b") int i = 0; i < a.length; i++) ;
+    }
+}

--- a/checker/tests/index/SameLenOnFormalParameter.java
+++ b/checker/tests/index/SameLenOnFormalParameter.java
@@ -1,30 +1,28 @@
 // Test case for issue #2434: http://tinyurl.com/cfissue/2434
 
-// @skip-test until the bug is fixed
-
 import org.checkerframework.checker.index.qual.*;
 
 class SameLenOnFormalParameter {
-    public void requiresSameLen1(String x, @SameLen("#1") String y) {}
+    public void requiresSameLen1(String x1, @SameLen("#1") String y1) {}
 
-    public void requiresSameLen2(@SameLen("#2") String x, String y) {}
+    public void requiresSameLen2(@SameLen("#2") String x2, String y2) {}
 
-    public void m1(@SameLen("#2") String a, String b) {
-        requiresSameLen1(a, b);
+    public void m1(@SameLen("#2") String a1, String b1) {
+        requiresSameLen1(a1, b1);
     }
 
-    public void m2(@SameLen("#2") String a, String b) {
-        @SameLen("a") String b2 = b;
-        requiresSameLen1(a, b2);
+    public void m2(@SameLen("#2") String a2, String b2) {
+        @SameLen("a2") String b22 = b2;
+        requiresSameLen1(a2, b22);
     }
 
-    public void m3(@SameLen("#2") String a, String b) {
-        @SameLen("b") String a2 = a;
-        @SameLen("a") String b2 = b;
-        requiresSameLen1(a, b2);
+    public void m3(@SameLen("#2") String a3, String b3) {
+        @SameLen("b3") String a2 = a3;
+        @SameLen("a3") String b32 = b3;
+        requiresSameLen1(a3, b32);
     }
 
-    public void m4(@SameLen("#2") String a, String b) {
-        requiresSameLen2(a, b);
+    public void m4(@SameLen("#2") String a4, String b4) {
+        requiresSameLen2(a4, b4);
     }
 }

--- a/checker/tests/index/SameLenOnFormalParameterSimple.java
+++ b/checker/tests/index/SameLenOnFormalParameterSimple.java
@@ -1,0 +1,11 @@
+// Test case for issue #2434: http://tinyurl.com/cfissue/2434
+
+import org.checkerframework.checker.index.qual.SameLen;
+
+class SameLenOnFormalParameterSimple {
+    public void requiresSameLen1(String x1, @SameLen("#1") String y1) {}
+
+    public void m1(@SameLen("#2") String a1, String b1) {
+        requiresSameLen1(a1, b1);
+    }
+}

